remove hard limit on the number of items returned when an index is queried

ezra-varady opened the pull request:
This should address #74 It includes a regression test where `...LIMIT 10` (the new default) and `...LIMIT 50` are tested.

Resolves #74

Ngalstyan4 commented on test/sql/hnsw_index_limit.sql:
```diff
@@ -0,0 +1,12 @@
+);
+
+\copy sift_base10k (v) FROM '/tmp/lanterndb/vector_datasets/siftsmall_base_arrays.csv' with csv;
+-- build index on an existing table of 10k rows
```
please add these tests to hnsw.sql instead of creating a new test file

Ngalstyan4 commented on test/sql/hnsw_index_limit.sql:
```diff
@@ -0,0 +1,12 @@
+-- build index on an existing table of 10k rows
+CREATE INDEX hnsw_idx ON sift_base10k USING hnsw (v) WITH (M=2, ef_construction=10, ef=4, dims=128);
+SELECT V AS v4444  FROM sift_base10k WHERE id = 4444 \gset
+SELECT * FROM sift_base10k order by v <-> :'v4444' LIMIT 10;
```
Can you do `SELECT count(*) ...` instead? we really are only testing the count here so lets make the `.out` file less noisy

Ngalstyan4 commented on test/sql/hnsw_index_limit.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE INDEX hnsw_idx ON sift_base10k USING hnsw (v) WITH (M=2, ef_construction=10, ef=4, dims=128);
+SELECT V AS v4444  FROM sift_base10k WHERE id = 4444 \gset
+SELECT * FROM sift_base10k order by v <-> :'v4444' LIMIT 10;
+SELECT * FROM sift_base10k order by v <-> :'v4444' LIMIT 50;
```
Can you think of more edge cases to test here?
E.g. What happens when I am asking for 50 but the table has 30 elements?

Ngalstyan4 replied:
Can you also test the case when the `init_k` variable is changed?
You may need to add a debug log every time `usearch_search` is called to notice the difference in sql output.

Ngalstyan4 commented on src/hnsw/scan.c:
```diff
@@ -211,6 +213,31 @@ bool ldb_amgettuple(IndexScanDesc scan, ScanDirection dir)
+        vec = DatumGetVector(value);
+
+        /* double k and reallocate arrays to account for increased size */
+        k = k * 2;
```
So you are using the static `k` to know how many items to retrieve next.
Could this be `k = scanstate->count * 2`?

It is very strange to see static variables in this context. I am pretty sure statics will do the wrong thing if two scans happen at the same time. So, if possible, let's get rid of the static variable.

ezra-varady replied:
good catch, this was poorly thought out

Ngalstyan4 reviewed (commented):
Thanks for your contribution, @ezra-varady !
This is great! And is almost ready. See comments below.

ezra-varady commented:
to summarize, 

- I fixed the issue with the static variable. You're absolutely right about it breaking when two queries occur simultaneously. An embarrassing mistake. 
- Moved new tests into hnsw.sql and made corresponding updates to the expected output
- Added debug logging of `init_k` when `usearch_search` is called, as well as tests that modify it 

I still need to add a test for the query being bigger than the table

Ngalstyan4 commented:
Thanks for the prompt fixes, @ezra-varady! 

> I still need to add a test for the query being bigger than the table. 

I am not sure what you want to test here. But if the problem is that the table has more rows than you need, you can place the test in a transaction, drop all the extra rows in the transaction, then run the query, then rollback the transaction.
See an example of this kind of thing [here](https://github.com/lanterndata/lanterndb/blob/c236c4e9d1957da9f97f9809819568da961a0cc1/test/sql/hnsw.sql#L81C6-L95).

Note that you also do not need a large table to force index usage in tests. you can just [disable sequential scan](https://github.com/lanterndata/lanterndb/blob/c236c4e9d1957da9f97f9809819568da961a0cc1/test/sql/hnsw_insert.sql#L7).

ezra-varady commented:
Thanks, I didn't know that! I reformatted the tests in light of this. I also added a check to avoid a potential extra query I had neglected. I think I had misunderstood your suggestion w.r.t. the additional test. Please let me know if you have other suggestions

Ngalstyan4 commented on src/hnsw/scan.c:
```diff
@@ -153,6 +153,7 @@ bool ldb_amgettuple(IndexScanDesc scan, ScanDirection dir)
 {
     HnswScanState *scanstate = (HnswScanState *)scan->opaque;
     ItemPointer    tid;
+    // todo:: fix me. if there is way to know how many we need, use that
```
unneeded comment or comment in the wrong place

Ngalstyan4 commented on src/hnsw/scan.c:
```diff
@@ -191,13 +190,13 @@ bool ldb_amgettuple(IndexScanDesc scan, ScanDirection dir)
         }
 
         // hnsw_search(scanstate->hnsw, vec->x, k, &num_returned, scanstate->distances, scanstate->labels);
+        elog(INFO, "querying index for %d elements", k);
```
Can you make sure this is only triggered for the tests you wrote and not for all the other tests as well?

Ngalstyan4 reviewed (approved):
lgtm!

One small thing:
To avoid the extra logs added to the rest of the regression tests due to the elog here, can you surround the tests that concern the dynamic k-value growth with the following statements?

```
SET client_min_messages TO DEBUG5;
--- YOUR TESTS HERE ---
RESET client_min_messages;
```
Then, if you change your added `elog(INFO,..` code lines to `elog(DEBUG5`, the log message will be printed only in the tests you needed.

As a side note, we should go do this for the rest of the regression tests as well.

ezra-varady commented:
removed the hanging comment and changed the log level for the tests
